warehouse/accounts/services.py

```diff
@@ -759,3 +772,163 @@ def check_password(self, password, *, tags=None):
         # This service allows *every* password as a non-breached password. It will never
         # tell a user their password isn't good enough.
         return False
+
+
+class GitHubPublicKeyMetaAPIError(InvalidTokenLeakRequest):
+    pass
+
+
+@implementer(IGitHubTokenScanningPayloadVerifyService)
+class GitHubTokenScanningPayloadVerifyService:
+    """
+    Checks payload signature using:
+    - `requests` for HTTP calls
+    - `cryptography` for signature verification
+    """
+
+    def __init__(self, *, session, metrics, api_token=None):
+        self._metrics = metrics
+        self._session = session
+        self._api_token = api_token
+
+    @classmethod
+    def create_service(cls, context, request):
+        return cls(
+            session=request.http,
+            metrics=request.find_service(IMetricsService, context=context),
+            api_token=request.registry.settings["github.token"],
+        )
+
+    def verify(self, *, payload, key_id, signature):
+
+        try:
+            pubkey_api_data = self._retrieve_public_key_payload()
+            public_keys = self._extract_public_keys(pubkey_api_data)
+            public_key = self._check_public_key(
+                github_public_keys=public_keys, key_id=key_id
+            )
+            self._check_signature(
+                payload=payload, public_key=public_key, signature=signature
+            )
+        except InvalidTokenLeakRequest as exc:
+            self._metrics.increment(
+                f"warehouse.token_leak.github.auth.error.{exc.reason}"
+            )
+            return False
+
+        self._metrics.increment("warehouse.token_leak.github.auth.success")
+        return True
+
+    def _get_headers(self):
+        if self._api_token:
+            return {"Authorization": f"token {self._api_token}"}
+        return {}
```

Comment:
As far as I know, without PAT, GitHub does rate-limiting based on IP, which means depending o how we deploy and what our externally seen IP is, we may be rate-limited du to calls from other people on the same datacenter.

Using a PAT is better, but it's not required. For example, if someone tries to use this code from a local environment, it's perfectly ok if they decide not to use PAT. This is what led me to having those 2 branches. I can add a comment if you want :)

Reply:
I guess I'm not sure what we'd want to do in production: it's my understanding that PATs are user-specific and can't be generated for organizations, so whose PAT should we use in prod?

Reply:
Hm, I’ll admit I’m not really sure. In my experience, organisations always had dedicated users for that, but it’s more of a habit than a rule of course. Also, I was under the impression that it was possible to use oauth apps for that but I didn’t investigate.

If it turns out it has to be someone’s token and there’s no « bot » Pypi user, and you wish not to create one, I guess all that’s left is no token. I’ve been bitten by the rate limit earlier so I tend to be mindful of that now, but if you think this won’t be a problem, I’m perfectly oK to drop it, as long as we’re aware of this, and if it breaks someday, we will know why.

Reply:
Maybe this is something we should talk to GitHub about what they expect users to do here? (I'm not on the original thread).

Reply:
I've asked github by e-mail, but I've tested in parallel, and unauthenticated calls, including to this URL, are limited to 60 calls / hour / IP. I think we can say this is clearly not enough. We'll need to cache the response and/or to make authenticated calls, but I don't think 60 calls/hour from the IP will be enough.

Reply:
Ok, it's been 5 days and no answer so far. Let's assume token is important, according to field research, and make it mandatory. It will always be time to change our minds later.

And for how we'll get a token for production: the more I think about it, the more I think a dedicated account is probably the easiest way to go. The token doesn't even need special access to anything: it only accesses an open resource.